pkg/operator/controller/controller_cabundle_configmap.go

```diff
@@ -0,0 +1,193 @@
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	cmLabels = map[string]string{
+		"dns.operator.openshift.io/ca-bundle": "true",
+	}
+	cmLabelSelector = metav1.LabelSelector{
+		MatchLabels: cmLabels,
+	}
+)
+
+// ensureCABundleConfigMaps syncs CA bundle configmaps for a DNS
+// between the openshift-config and openshift-dns namespaces if the user has
+// configured a CA bundle configmap.  Returns a Boolean indicating whether the
+// configmap exists, the configmap if it does exist, and an error value.
+func (r *reconciler) ensureCABundleConfigMaps(dns *operatorv1.DNS) error {
+	var configmapNames []string
+	transportConfig := dns.Spec.UpstreamResolvers.TransportConfig
+	if transportConfig.Transport == operatorv1.TLSTransport && transportConfig.TLS != nil && transportConfig.TLS.CABundle.Name != "" {
+		configmapNames = append(configmapNames, transportConfig.TLS.CABundle.Name)
+	}
+	for _, server := range dns.Spec.Servers {
+		transportConfig := server.ForwardPlugin.TransportConfig
+		if transportConfig.Transport == operatorv1.TLSTransport && transportConfig.TLS != nil && transportConfig.TLS.CABundle.Name != "" {
+			configmapNames = append(configmapNames, transportConfig.TLS.CABundle.Name)
+		}
+	}
+
+	for _, name := range configmapNames {
+		sourceName := types.NamespacedName{
+			Namespace: GlobalUserSpecifiedConfigNamespace,
+			Name:      name,
+		}
+		haveSource, source, err := r.currentCABundleConfigMap(sourceName)
+		if err != nil {
+			return err
+		}
+
+		destName := CABundleConfigMapName(source.Name)
+		have, current, err := r.currentCABundleConfigMap(destName)
+		if err != nil {
+			return err
+		}
+
+		want, desired, err := desiredCABundleConfigMap(dns, haveSource, source, destName)
+		if err != nil {
+			return err
+		}
+
+		switch {
+		case !want && !have:
+			return nil
+		case !want && have:
+			if err := r.client.Delete(context.TODO(), current); err != nil {
+				if !errors.IsNotFound(err) {
+					return fmt.Errorf("failed to delete configmap: %w", err)
+				}
+			} else {
+				logrus.Infof("deleted configmap %s/%s", current.Namespace, current.Name)
+			}
+			return nil
+		case want && !have:
+			if err := r.client.Create(context.TODO(), desired); err != nil {
+				return fmt.Errorf("failed to create configmap: %w", err)
+			}
+			logrus.Infof("created configmap %s/%s", desired.Namespace, desired.Name)
+			_, _, err = r.currentCABundleConfigMap(destName)
+			return err
+		case want && have:
+			if updated, err := r.updateCABundleConfigMap(current, desired); err != nil {
+				return fmt.Errorf("failed to update configmap: %w", err)
+			} else if updated {
+				logrus.Infof("updated configmap %s/%s", desired.Namespace, desired.Name)
+				_, _, err = r.currentCABundleConfigMap(destName)
+				return err
+			}
+		}
+	}
+
+	// remove ca bundle configmaps that are not referred in dns anymore.
+	cmSelector, err := metav1.LabelSelectorAsSelector(&cmLabelSelector)
```

Comment:
The label selector doesn't change, so you could build `cmSelector` just once:
```go
var (
	cmLabels = map[string]string{
		"dns.operator.openshift.io/ca-bundle": "true",
	}
	cmLabelSelector = metav1.LabelSelector{
		MatchLabels: cmLabels,
	}
	// cmSelector is a labels.Selector built from cmLabelSelector.
	cmSelector = func() labels.Selector {
		v, err := metav1.LabelSelectorAsSelector(&cmLabelSelector)
		if err != nil {
			panic(err)
		}
		return v
	}()
)
```
(You'll need to import `"k8s.io/apimachinery/pkg/labels"`.)

Reply:
Done.